File: e2e/liveness.ts
// Liveness probes for the soak lane (#1603, plan #1600 §3 Phase 4.1).
//
// Everything in this file exists to answer one question the rest of this
// repo's guards structurally cannot (plan #1600 §2.2): **after the app has
// been running for a while under load, does a keystroke still reach a pane
// and does the MCP still answer?** Four hangs — #1564, #1592, #1595 and the
// beta6 field report — shipped past a growing wall of source scans, because a
// source scan pins the shape of the last incident and this is a liveness
// property.
//
// Three design choices here are load-bearing rather than incidental:
//
// - **Every probe is bounded by its own deadline, not by Playwright's test
//   timeout.** The failure under test is a HANG. A probe that simply awaited
//   would turn every red into "Test timeout of 120000ms exceeded" with no
//   statement of which half died, and would make the expected-failure marker
//   on the class assertion unreliable. Each probe returns `{ ok, ms }` and the
//   spec asserts on that, so a hang reads as a named assertion failure with a
//   measured elapsed time.
// - **The MCP probe goes over real HTTP from the Playwright process**, not
//   through the webview. The webview is the half of the app that stays alive
//   in the beta6 mechanism (its thread is untouched — that is why the window
//   still paints), so asking it whether the backend is well is asking the
//   wrong process. `tiny_http` + `std::thread::spawn` per request is what the
//   orchestrator actually talks to, so that is what gets probed.
// - **Nothing here spawns an agent CLI** (CLAUDE.md constraint 3). The MCP
//   token comes from `orch_solo_prepare`, which mints a token and writes the
//   config file *before* any pane process is spawned and independently of
//   whether the CLI is even installed — so calling that command alone yields
//   a working MCP identity with no child process anywhere.
import * as fs from "node:fs";
import * as path from "node:path";
import { type Locator, type Page } from "@playwright/test";

/** Reads a soak knob, accepting both brand spellings. Emit sites use
 *  `ORRERIX_`; a reader keeps every accepted spelling
 *  (doc/design/rebrand-protocol.md). */
function envInt(suffix: string, fallback: number): number {
  const raw = process.env[`ORRERIX_${suffix}`] ?? process.env[`LOOMUX_${suffix}`];
  if (raw === undefined || raw.trim() === "") return fallback;
  const n = Number(raw);
  if (!Number.isFinite(n) || n < 0) {
    throw new Error(`ORRERIX_${suffix}=${raw} is not a non-negative number`);
  }
  return Math.floor(n);
}

// ---------------------------------------------------------------------------
// Budget. Every number here is an env knob, because the CI default and a long
// local soak want different answers and the whole point of a soak lane is that
// somebody can turn it up. See doc/design/e2e-testing.md for the CI budget.
// ---------------------------------------------------------------------------

/** How long the app idles under poll load before anything is asserted.
 *  Default 180 s: `src/tabbar.ts`'s status poll is 4 s, so that is ~45 ticks
 *  per bound tab — many dozens of poll invokes in total, which is the scale
 *  the plan asks for, at a cost this job can afford on every PR. */
export const SOAK_MS = envInt("SOAK_MS", 180_000);

/** How long the injected registry-lock hold lasts. Default 30 s: long enough
 *  to span the two liveness probes several times over. Raising this past
 *  ~120 s is what a local run does to reach blocking-pool exhaustion, which is
 *  the second half of the beta6 mechanism (plan #1600 §1.2 step 4). */
export const LOCK_HOLD_MS = envInt("SOAK_LOCK_HOLD_MS", 30_000);

/** The bound every liveness probe must answer within. */
export const LIVENESS_BOUND_MS = envInt("SOAK_BOUND_MS", 8_000);

/** Corpus size. Defaults chosen so the fixture build stays under a couple of
 *  seconds while still being the "large corpus" shape #1592 was reported
 *  against. */
export const CORPUS_GROUPS = envInt("SOAK_GROUPS", 8);
export const CORPUS_SESSIONS = envInt("SOAK_SESSIONS", 900);
export const CORPUS_AUDIT_LINES = envInt("SOAK_AUDIT_LINES", 4_000);

// ---------------------------------------------------------------------------
// The lock-hold injector's file protocol. These three literals are the joint
// with src-tauri/src/orchestration/e2ehold.rs, and
// src-tauri/tests/e2ehold_guard.rs asserts this file still mentions all three
// — a rename on either side otherwise produces a soak run with no hold behind
// it, which is green and meaningless.
// ---------------------------------------------------------------------------

export const HOLD_ENV_VAR = "ORRERIX_E2E_LOCK_HOLD";
export const HOLD_REQUEST_FILE = "e2e-lock-hold.request";
export const HOLD_STATE_FILE = "e2e-lock-hold.state";

export type HoldTarget = "groups" | "agents" | "by_token";

export interface HoldState {
  target?: string;
  hold_ms?: number;
  requested_ms?: number;
  acquired_ms?: number | null;
  released_ms?: number | null;
  error?: string;
}

/** Asks the app under test to take a registry mutex and hold it. Returns as
 *  soon as the request is on disk — `waitForHoldAcquired` is what confirms it
 *  was honoured. */
export function requestLockHold(dataDir: string, target: HoldTarget, holdMs: number): void {
  fs.writeFileSync(
    path.join(dataDir, HOLD_REQUEST_FILE),
    JSON.stringify({ target, hold_ms: holdMs }),
    "utf8"
  );
}

/** Reads the injector's breadcrumb, tolerating "not written yet". The writer
 *  renames into place, so a torn read should not be possible — a malformed
 *  parse is still treated as "not yet" rather than thrown, because the only
 *  thing a throw here would achieve is turning a timing wobble into a red on
 *  a spec about something else. */
export function readHoldState(dataDir: string): HoldState | null {
  try {
    return JSON.parse(fs.readFileSync(path.join(dataDir, HOLD_STATE_FILE), "utf8")) as HoldState;
  } catch {
    return null;
  }
}

/** Blocks until the injector reports it has ACQUIRED the mutex.
 *
 *  This is the positive control for the whole class assertion. A hold that
 *  silently never happened is indistinguishable, from the assertion's side,
 *  from the app surviving one — so the spec must never run its bounded probes
 *  without first having seen `acquired_ms`. */
export async function waitForHoldAcquired(
  dataDir: string,
  timeoutMs = 15_000
): Promise<HoldState> {
  const deadline = Date.now() + timeoutMs;
  let last: HoldState | null = null;
  while (Date.now() < deadline) {
    last = readHoldState(dataDir);
    if (last?.error) {
      throw new Error(`the lock-hold injector refused the request: ${last.error}`);
    }
    if (typeof last?.acquired_ms === "number" && last.acquired_ms > 0) return last;
    await sleep(100);
  }
  throw new Error(
    `the lock-hold injector never reported acquiring a lock within ${timeoutMs}ms ` +
      `(last state: ${JSON.stringify(last)}). Either the app was not launched with ` +
      `${HOLD_ENV_VAR}=1, or it is a release build, where the injector does not exist.`
  );
}

/** Whether the hold is still in effect. Read AFTER the liveness probes: a
 *  hold that expired before they ran would leave them measuring an idle app. */
export function holdStillHeld(dataDir: string): boolean {
  const s = readHoldState(dataDir);
  return typeof s?.acquired_ms === "number" && (s.released_ms ?? null) === null;
}

export function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

// ---------------------------------------------------------------------------
// Invoke accounting — the positive control for the soak itself.
// ---------------------------------------------------------------------------

/** Counts every command the frontend dispatches, by name.
 *
 *  Without this the soak is an assertion about a load nobody measured: if the
 *  tab-strip poll were disarmed, or the corpus's tabs failed to bind, the app
 *  would idle for three minutes doing nothing and the liveness probes would
 *  pass — a green run that proves the app survives being left alone.
 *
 *  `@tauri-apps/api`'s `invoke` reads `window.__TAURI_INTERNALS__.invoke` at
 *  call time (node_modules/@tauri-apps/api/core.js), so wrapping the property
 *  intercepts every dispatch the real frontend makes, through
 *  `src/transport.ts` and every bridge above it. */
export async function installInvokeCounter(page: Page): Promise<void> {
  await page.evaluate(() => {
    const w = window as unknown as {
      __TAURI_INTERNALS__?: {
        invoke(cmd: string, args?: unknown, options?: unknown): Promise<unknown>;
        __soakOriginalInvoke?: (cmd: string, args?: unknown, options?: unknown) => Promise<unknown>;
      };
      __soakInvokeCounts?: Record<string, number>;
    };
    const internals = w.__TAURI_INTERNALS__;
    if (!internals) throw new Error("__TAURI_INTERNALS__ missing — not running inside the app");
    if (internals.__soakOriginalInvoke) return;
    internals.__soakOriginalInvoke = internals.invoke.bind(internals);
    w.__soakInvokeCounts = {};
    internals.invoke = (cmd: string, args?: unknown, options?: unknown) => {
      const counts = w.__soakInvokeCounts as Record<string, number>;
      counts[cmd] = (counts[cmd] ?? 0) + 1;
      return internals.__soakOriginalInvoke!(cmd, args, options);
    };
  });
}

export async function readInvokeCounts(page: Page): Promise<Record<string, number>> {
  return await page.evaluate(
    () => ({ ...((window as unknown as { __soakInvokeCounts?: Record<string, number> }).__soakInvokeCounts ?? {}) })
  );
}

/** Total dispatches whose command name starts with `orch_` — the poll load,
 *  independent of which particular commands a given release polls (the
 *  group-view batch has been five, then nine, then ten; pinning a count here
 *  would pin the last incident's shape, which is the mistake this whole lane
 *  exists to stop making). */
export function orchInvokeTotal(counts: Record<string, number>): number {
  return Object.entries(counts)
    .filter(([cmd]) => cmd.startsWith("orch_"))
    .reduce((sum, [, n]) => sum + n, 0);
}

// ---------------------------------------------------------------------------
// Probe A — a keystroke reaches a pane, and the child's output comes back.
// ---------------------------------------------------------------------------

/** Taps the backend's `pty-output` event stream from the page.
 *
 *  xterm.js renders through the WebGL addon, so a terminal's contents are NOT
 *  in the DOM and cannot be read with a selector. Rather than add a debug hook
 *  to product code to expose the buffer, this listens to the same Tauri event
 *  the app's own router consumes (`src/pty.ts`'s `listen<{id,data}>
 *  ("pty-output")`), through the low-level bridge the bundled `listen()` uses
 *  underneath — the same technique `queue-badge.spec.ts` already uses in the
 *  emit direction, and permitted by the shipped ACL (`core:default` covers
 *  `core:event`). */
export async function installPtyTap(page: Page): Promise<void> {
  await page.evaluate(async () => {
    const w = window as unknown as {
      __TAURI_INTERNALS__?: {
        invoke(cmd: string, args?: unknown): Promise<unknown>;
        transformCallback(cb: (e: unknown) => void, once?: boolean): number;
      };
      __soakPtyText?: string;
    };
    const internals = w.__TAURI_INTERNALS__;
    if (!internals) throw new Error("__TAURI_INTERNALS__ missing — not running inside the app");
    if (typeof w.__soakPtyText === "string") return;
    w.__soakPtyText = "";
    const handler = internals.transformCallback((raw: unknown) => {
      const payload = (raw as { payload?: { data?: unknown } } | null)?.payload;
      if (payload && typeof payload.data === "string") {
        // Bounded: a soak can run for an hour locally, and an unbounded string
        // in the page would be this harness's own INV-8 violation.
        const next = (w.__soakPtyText ?? "") + payload.data;
        w.__soakPtyText = next.length > 65_536 ? next.slice(next.length - 65_536) : next;
      }
    });
    await internals.invoke("plugin:event|listen", {
      event: "pty-output",
      target: { kind: "Any" },
      handler,
    });
  });
}

export async function ptyTapText(page: Page): Promise<string> {
  return await page.evaluate(
    () => (window as unknown as { __soakPtyText?: string }).__soakPtyText ?? ""
  );
}

/** ANSI/control-sequence stripped view of the tap, so a shell's cursor
 *  positioning and colour codes do not sit between the characters being
 *  matched. Three families, because a Windows console emits all three: OSC
 *  (title sets, terminated by BEL or ST), CSI (the `ESC [` majority), and
 *  the bare two-character escapes. */
export function stripAnsi(text: string): string {
  return text
    .replace(/\u001b\][\s\S]*?(?:\u0007|\u001b\\)/g, "")
    .replace(/\u001b\[[0-9;?]*[ -/]*[@-~]/g, "")
    .replace(/\u001b[@-Z\\-_]/g, "");
}

export interface ProbeResult {
  ok: boolean;
  ms: number;
  detail: string;
}

/**
 * Types a command into the focused pane and waits for the CHILD's own output
 * to come back — bounded by `boundMs`, never by the test timeout.
 *
 * The command is `echo soak<n>_%RANDOM%_end` and the match is
 * `/soak<n>_\d+_end/`. That asymmetry is deliberate and is what makes this a
 * round trip rather than a terminal echo: the text typed in contains the
 * literal `%RANDOM%`, so a match on the digit form can only have been produced
 * by `cmd.exe` expanding it. A pane that rendered the keystrokes locally but
 * never reached — or never heard back from — its child cannot produce it.
 *
 * This is the assertion the beta6 report is about: `src/ptywrite.ts` keeps one
 * `write_pty` in flight per pane and chains the next on the previous promise
 * (#65), so a `write_pty` that never resolves stops that pane accepting input
 * forever, with the window still painting.
 */
export async function ptyRoundTrip(
  page: Page,
  paneTerm: Locator,
  nonce: number,
  boundMs: number
): Promise<ProbeResult> {
  const marker = `soak${nonce}`;
  const expected = new RegExp(`${marker}_\\d+_end`);
  const started = Date.now();

  await paneTerm.click();
  await page.keyboard.type(`echo ${marker}_%RANDOM%_end`);
  await page.keyboard.press("Enter");

  while (Date.now() - started < boundMs) {
    const text = stripAnsi(await ptyTapText(page));
    if (expected.test(text)) {
      return { ok: true, ms: Date.now() - started, detail: `matched ${expected}` };
    }
    await sleep(150);
  }
  const tail = stripAnsi(await ptyTapText(page)).slice(-400);
  return {
    ok: false,
    ms: Date.now() - started,
    detail: `no output matching ${expected} within ${boundMs}ms; last 400 chars of the pty tap: ${JSON.stringify(tail)}`,
  };
}

// ---------------------------------------------------------------------------
// Probe B — the MCP still answers.
// ---------------------------------------------------------------------------

export interface McpEndpoint {
  url: string;
  header: string;
  token: string;
  agentId: string;
}

/**
 * Mints a real MCP identity without spawning anything.
 *
 * `orch_solo_prepare` mints a token, registers an `AgentEntry`, and writes the
 * agent's MCP config file — all before the launcher would spawn the pane's
 * process, and independently of whether that CLI is installed. Calling it
 * directly therefore yields a valid token and the server's ephemeral port
 * (which is bound at `127.0.0.1:0` in `mcp::serve` and is not otherwise
 * discoverable) with no child process anywhere. CLAUDE.md constraint 3 is
 * about spawning agent CLIs; this spawns nothing.
 *
 * The config's exact JSON layout differs per CLI, so the url/header pair is
 * located by SHAPE — an object carrying a `/mcp` url and a headers map —
 * rather than by a hard-coded key path that would break silently on the next
 * CLI added.
 */
export async function mintMcpEndpoint(
  page: Page,
  dataDir: string,
  cwd: string
): Promise<McpEndpoint> {
  const prepared = (await page.evaluate(async (paneCwd: string) => {
    const internals = (
      window as unknown as {
        __TAURI_INTERNALS__?: { invoke(cmd: string, args: unknown): Promise<unknown> };
      }
    ).__TAURI_INTERNALS__;
    if (!internals) throw new Error("__TAURI_INTERNALS__ missing — not running inside the app");
    return (await internals.invoke("orch_solo_prepare", {
      cli: "copilot",
      cwd: paneCwd,
      name: "soak-probe",
    })) as { agent_id?: string; delivery_only?: boolean };
  }, cwd)) as { agent_id?: string; delivery_only?: boolean };

  const agentId = prepared?.agent_id;
  if (!agentId) {
    throw new Error(`orch_solo_prepare returned no agent_id: ${JSON.stringify(prepared)}`);
  }
  if (prepared.delivery_only) {
    throw new Error(
      `orch_solo_prepare reported delivery_only for this CLI — no token was minted, so there ` +
        `is no MCP identity to probe with`
    );
  }

  const cfgPath = path.join(dataDir, "orchestration", "__solo__", "configs", `${agentId}.json`);
  const cfg = JSON.parse(fs.readFileSync(cfgPath, "utf8")) as unknown;
  const found = findServerEntry(cfg);
  if (!found) {
    throw new Error(`no MCP server entry with a url and headers found in ${cfgPath}`);
  }
  return { ...found, agentId };
}

function findServerEntry(node: unknown): { url: string; header: string; token: string } | null {
  if (node === null || typeof node !== "object") return null;
  const obj = node as Record<string, unknown>;
  const url = obj.url;
  const headers = obj.headers;
  if (typeof url === "string" && url.includes("/mcp") && headers && typeof headers === "object") {
    for (const [header, token] of Object.entries(headers as Record<string, unknown>)) {
      if (typeof token === "string" && header.toLowerCase().endsWith("-agent")) {
        return { url, header, token };
      }
    }
  }
  for (const child of Object.values(obj)) {
    const hit = findServerEntry(child);
    if (hit) return hit;
  }
  return null;
}

export interface McpResult extends ProbeResult {
  status?: number;
  body?: unknown;
}

/**
 * One JSON-RPC call against the running app's MCP server, bounded.
 *
 * `ping` is the cheapest method and still the faithful probe: every method,
 * `ping` included, is authenticated first, and `resolve_token` locks
 * `by_token`, then `agents`, then `groups`. So a `ping` that answers is
 * evidence the registry is reachable, and a `ping` that does not is the exact
 * first symptom of the beta6 mechanism — "MCP dies first" (plan #1600 §1.2
 * step 2).
 */
export async function mcpCall(
  ep: McpEndpoint,
  method: string,
  params: unknown,
  boundMs: number,
  tokenOverride?: string
): Promise<McpResult> {
  const started = Date.now();
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), boundMs);
  try {
    const res = await fetch(ep.url, {
      method: "POST",
      headers: { "content-type": "application/json", [ep.header]: tokenOverride ?? ep.token },
      body: JSON.stringify({ jsonrpc: "2.0", id: 1, method, params }),
      signal: controller.signal,
    });
    const body = (await res.json()) as unknown;
    return {
      ok: true,
      ms: Date.now() - started,
      status: res.status,
      body,
      detail: `${method} → HTTP ${res.status}`,
    };
  } catch (err) {
    return {
      ok: false,
      ms: Date.now() - started,
      detail: `${method} did not answer within ${boundMs}ms: ${String(err)}`,
    };
  } finally {
    clearTimeout(timer);
  }
}

/** Whether an `McpResult` is a JSON-RPC SUCCESS (not an error envelope — the
 *  server returns those with HTTP 200 too, so status alone decides nothing). */
export function isJsonRpcResult(r: McpResult): boolean {
  const b = r.body as { jsonrpc?: unknown; result?: unknown } | undefined;
  return r.ok && b?.jsonrpc === "2.0" && b.result !== undefined;
}

/** Whether an `McpResult` is a JSON-RPC ERROR envelope. Used for the negative
 *  control: a bogus token must be REFUSED, which is what distinguishes "we are
 *  talking to orrerix's authenticated MCP server" from "something answered on
 *  that port". */
export function jsonRpcErrorCode(r: McpResult): number | null {
  const b = r.body as { error?: { code?: unknown } } | undefined;
  return typeof b?.error?.code === "number" ? b.error.code : null;
}

File: src-tauri/tests/liveness.rs
//! Liveness tests: does the app still answer while a registry lock is held?
//!
//! Plan #1600 §2.2 is why this file exists. Every other guard in this repo is a
//! **shape** scan — `perf_dispatch.rs` reads source and asserts structure — and
//! all four hangs (#1564, #1592, #1595 and the beta6 field report) shipped past
//! a growing wall of them, because a shape scan pins the last incident and none
//! of them can ask *does this still return while something is stuck*. These
//! ask it.
//!
//! Must be an integration test, not a unit test (CLAUDE.md constraint 4 — the
//! Windows test exe needs build.rs's comctl32-v6 manifest). No real agent CLI
//! anywhere (constraint 3): the holds come from
//! `OrchRegistry::hold_lock_for_test`, Phase 0's `#[doc(hidden)]` seam.
//!
//! | id | property | slice |
//! |---|---|---|
//! | L1 | both published reads return while every holdable lock is held; `stale` flips on the clock and clears on a publish | Phase 1 (#1608) |
//!
//! L0 (the negative control as a test of its own) and L3 (the pty writer's pool
//! isolation, with its pool-depth clause) belong to #1612 and land with it;
//! whichever of the two PRs is second APPENDS rather than rewrites. L1 keeps
//! its own in-test control so it is non-vacuous either side of that merge.

use std::sync::{mpsc, Arc};
use std::time::{Duration, Instant};

use serde_json::Value;

use loomux_engine::lockwatch::tracked_lock_names;
use loomux_lib::orchestration::views::{group_view_payload, strip_view_payload, VIEW_STALE_AFTER_MS};
use loomux_lib::orchestration::{Guardrails, OrchRegistry};

/// Long enough that a loaded CI runner never trips it, short enough that a real
/// regression fails the job rather than hanging it. Every use is a "did this
/// make progress at all" question, not a latency measurement — the fix moves
/// the wait from unbounded to nothing, so there is no near-miss to tune around.
/// Same value and same reasoning as `perf_leaflocks.rs`.
const GRACE: Duration = Duration::from_secs(10);

/// How long a parked call is given to *be* parked before the property is
/// probed. Asserted as a NEGATIVE (`recv_timeout` must time out), so this is
/// not a guess about scheduling: a call that has not returned after 300 ms,
/// when its unparked cost is microseconds, is parked at the one place in it
/// that can block.
const SETTLE: Duration = Duration::from_millis(300);

/// Longer than any assertion below needs, so a hold never expires mid-probe and
/// turns a real failure into a flake that reads as a pass.
const HOLD_MS: u64 = 30_000;

fn rails() -> Guardrails {
    Guardrails { max_agents: 4, agent_cli: "claude".into(), ..Guardrails::default() }
}

/// #464: every registry construction here goes through this, so no spawn can
/// write into the developer's REAL `~/.claude/agents` or `~/.copilot/agents`.
fn test_registry() -> (Arc<OrchRegistry>, tempfile::TempDir) {
    let dir = tempfile::tempdir().unwrap();
    let reg = OrchRegistry::new(dir.path().to_path_buf());
    reg.set_port(45993);
    reg.set_claude_agents_dir_override(dir.path().join("claude-agents"));
    reg.set_copilot_agents_dir_override(dir.path().join("copilot-agents"));
    reg.set_compact_hook_dir_override(dir.path().join("compacthook"));
    reg.set_copilot_hooks_dir_override(dir.path().join("copilot-hooks"));
    (Arc::new(reg), dir)
}

/// Run `f` on its own thread; report whether it finished within `t`.
///
/// A `false` means the call is still blocked. The thread is left parked
/// deliberately: it is waiting on a lock this test still holds, and the harness
/// exits the process when the run ends.
fn completes_within<T: Send + 'static>(t: Duration, f: impl FnOnce() -> T + Send + 'static) -> bool {
    let (tx, rx) = mpsc::channel();
    std::thread::spawn(move || {
        let _ = tx.send(f());
    });
    rx.recv_timeout(t).is_ok()
}

// ---------- L1: the published reads answer regardless ----------

/// Which of `tracked_lock_names()` Phase 0's seam can actually hold.
///
/// `hold_lock_for_test` knows four names and returns `false` for the rest — a
/// deliberate choice documented at that seam ("a representative handful rather
/// than all 82"). So this test iterates EVERY tracked name, holds the ones it
/// can, and reports the rest as a stated residual rather than silently
/// covering four and reading as covering all of them. Widening the seam widens
/// this test with no edit here.
fn holdable_and_refused(reg: &Arc<OrchRegistry>) -> (Vec<String>, Vec<String>) {
    let mut names: Vec<String> = tracked_lock_names().into_iter().map(str::to_string).collect();
    // `tracked_lock_names` is a process-global registry of live locks, so a
    // second registry in this process would list every name twice.
    names.sort();
    names.dedup();

    let mut holdable = Vec::new();
    let mut refused = Vec::new();
    for name in names {
        // A 1 ms probe: this only asks whether the seam KNOWS the name. The
        // real holds are taken per-lock in the test below.
        if reg.hold_lock_for_test(&name, 1) {
            holdable.push(name);
        } else {
            refused.push(name);
        }
    }
    (holdable, refused)
}

#[test]
fn l1_a_published_read_returns_while_every_holdable_registry_lock_is_held() {
    let (reg, _d) = test_registry();
    let g = reg.create_group("C:/tmp/repo", rails()).unwrap();

    // Vacuity guard: the count `tracked_lock_names()` ACTUALLY returns. Phase 0
    // measured 82 (the plan said 85); this is a floor, not a pin, because the
    // number moves whenever a registry field is added and a test that has to be
    // edited for that is a test people edit without reading.
    let total = {
        let mut n: Vec<&str> = tracked_lock_names();
        n.sort_unstable();
        n.dedup();
        n.len()
    };
    assert!(
        total >= 80,
        "tracked_lock_names() returned only {total} names — the lock registry stopped \
         registering, so iterating it proves nothing"
    );

    let (holdable, refused) = holdable_and_refused(&reg);
    assert!(
        holdable.len() >= 4,
        "Phase 0's hold seam accepted only {} of {total} tracked locks ({holdable:?}) — it knows \
         four by name, so fewer than that means the seam broke, not that the registry shrank",
        holdable.len()
    );
    assert_eq!(
        holdable.len() + refused.len(),
        total,
        "every tracked lock must be classified as holdable or refused; the two lists must \
         partition the registry, or this test is quietly skipping some"
    );

    // THE PROPERTY. For each lock the seam can hold: both published reads must
    // return. Before #1608 each of these was ten (and two) registry
    // acquisitions per tick, so this is exactly L0's shape with the fix in.
    for name in &holdable {
        assert!(
            reg.hold_lock_for_test(name, HOLD_MS),
            "setup: the `{name}` hold must be real, or the assertions below prove nothing"
        );

        // NEGATIVE CONTROL, per lock, and the reason this test is not a
        // tautology: a REGISTRY read must NOT return while `agents` is held.
        // If the seam ever stops actually holding, every `completes_within`
        // below starts passing for the wrong reason and reads exactly like
        // coverage. Only `agents` is probed this way — it is the lock
        // `group_summary` takes first — so the control is asserted against a
        // read whose parking is known, not against every lock's shape.
        //
        // (`tests/liveness.rs` L0 states the same class as a test of its own;
        // it arrives with #1612. This is the in-test control, so L1 is
        // non-vacuous standing alone, before or after that lands.)
        if name == "agents" {
            let probe = reg.clone();
            let gid = g.id.clone();
            assert!(
                !completes_within(SETTLE, move || probe.group_summary(&gid)),
                "SETUP FAILURE, not a pass: `group_summary` returned while `agents` was held, \
                 so the hold is not holding and every assertion in this test is vacuous"
            );
        }

        let probe = reg.clone();
        let gid = g.id.clone();
        assert!(
            completes_within(GRACE, move || {
                group_view_payload(&probe.views.load(), &gid, Instant::now())
            }),
            "orch_group_view's body did not return while `{name}` was held. A polled read must \
             take NO registry lock: that is what stops one long hold parking a blocking-pool \
             thread per poller per tick until write_pty cannot be scheduled (#1600 §1.2)"
        );

        let probe = reg.clone();
        assert!(
            completes_within(GRACE, move || {
                strip_view_payload(&probe.views.load(), Instant::now())
            }),
            "orch_strip_view's body did not return while `{name}` was held (same property, the \
             other poll site)"
        );
    }

    // The residual, stated rather than implied — and printed, so a widening of
    // the seam is visible in the log rather than needing an edit here.
    println!(
        "L1 covered {}/{total} tracked locks: {holdable:?}. NOT holdable by Phase 0's seam \
         (stated residual, not a silent gap): {} names.",
        holdable.len(),
        refused.len()
    );
}

#[test]
fn l1_stale_flips_on_the_clock_while_a_lock_is_held_and_clears_on_the_next_publish() {
    let (reg, _d) = test_registry();
    let g = reg.create_group("C:/tmp/repo", rails()).unwrap();

    // One good publish, then the registry wedges.
    let published_at = Instant::now();
    reg.views.note_view_lease_at(&g.id, published_at);
    reg.views.publish_pass_at(&reg, published_at);
    assert!(reg.hold_lock_for_test("groups", HOLD_MS), "setup: the `groups` hold must be real");

    let stale_at = |now: Instant| -> bool {
        let payload = group_view_payload(&reg.views.load(), &g.id, now);
        payload
            .get("meta")
            .and_then(|m| m.get("stale"))
            .and_then(Value::as_bool)
            .expect("meta.stale is a bool")
    };

    // The read still answers with the wedge in place — that is L1 — and what it
    // answers is HONEST about its age. Injected clock rather than a sleep: the
    // property is the threshold, not the wall time.
    assert!(!stale_at(published_at), "a payload just published is not stale");
    assert!(
        stale_at(published_at + Duration::from_millis(VIEW_STALE_AFTER_MS + 1)),
        "past the threshold, a snapshot the publisher can no longer refresh must report stale — \
         a frozen panel that looks live is the disclosure gap #1604 review N3 deferred here"
    );
    assert!(
        stale_at(published_at + Duration::from_secs(3600)),
        "and waiting longer never clears it: the badge is released by EVIDENCE (the next \
         successful store), never by elapsed time"
    );

    // RELEASE ON EVIDENCE. The publisher is still parked on `groups`, so the
    // only way to clear the badge is a publish that really happens. Nothing
    // here waits for the hold: `publish_group_at` recomputes ONE group, and the
    // sections it needs for the strip tier are not behind `groups`.
    let recovered = Instant::now();
    assert!(
        completes_within(GRACE, {
            let reg = reg.clone();
            let gid = g.id.clone();
            move || reg.views.publish_group_at(&reg, &gid, recovered)
        }),
        "a single-group republish must not park behind `groups` either"
    );
    assert!(!stale_at(recovered), "one successful publish is the evidence that clears the badge");
}
